docs/spec.md

```diff
@@ -173,10 +188,48 @@ values of type `tensor<T>`).
   and an **imaginary part** of the same **element type**. Supported complex
   types are `complex<f32>` (both parts are of type `f32`) and `complex<f64>`
   (both parts are of type `f64`).
-* In the future, we are also planning to introduce **quantized types** that
-  represent integer values obtained via uniform quantization of floating-point
-  values using given scales and zero points
-  ([#588](https://github.com/openxla/stablehlo/issues/588)).
+* **Quantized types** represent integer values of a **storage type** in the
+  range from `storage_min` to `storage_max` (inclusive) that correspond to
+  floating-point values of an **expressed type**. For a given integer value `i`,
+  the corresponding floating-point value `f` can be computed as
+  `f = (i - zero_point) * scale`, where `scale` and `zero_point` are called
+  **quantization parameters**.
+
+Quantization can be **per-tensor**, meaning, having one `scale` and/or
+`zero_point` for the entire tensor or can be **per-axis**, meaning, having
+multiple `scales` and/or `zero_points`, one pair per slice of a particular
+dimension `quantized_dimension`. More formally, in a tensor `t` of per-axis
+quantized element type, there are `dim(t, quantized_dimension)` slices of the
+`quantized_dimension`: `t[:, ..., 0, ..., :], t[:, ..., 1, ..., :]`, etc. All
+elements in the `i`th slice use `scales[i]` and `zero_points[i]` as their
+quantization parameters.
+
+Quantized types satisfy the following constraints:
+
+* (C1) `num_bits(storage_type) < num_bits(expressed_type)`.
+* (C2) `type(storage_min) = storage_type`.
+* (C3) `type(storage_max) = storage_type`.
+* (C4) `storage_min` and `storage_max` have the following values:
+  * If `is_signed(storage_type)` and `d = num_bits(storage_type)`:
+    * `storage_min = -2^(d-1)`, `storage_max = 2^(d-1)-1` OR
+    * `storage_min = -2^(d-1)+1`, `storage_max = 2^(d-1)-1`.
+  * If `is_unsigned(storage_type)`:
+    * `storage_min = 0`, `storage_max = 2^d - 1`.
+* (C5) For all `i`, `type(scales[i]) = expressed_type`.
+* (C6) For all `i`, `scales[i] > 0`.
+* (C7) For all `i`, `type(zero_points[i]) = i64`.
```

Comment:
Thanks @dominicsymes for your valuable feedback.

Since `zero_point` can get out of the  storage value range, the proposal is using a wider type `i64` than the storage type, with max limit as [32](https://github.com/llvm/llvm-project/blob/0676c6d91fef302e0ae17f1a4571f28b1be1b27c/mlir/include/mlir/Dialect/Quant/QuantTypes.h#L58).


>> Additionally, I think zero points are most used for small data types – such 8-bit and below.

That is pretty useful information and can help in restricting the `zero_point` type. I will explore it further and welcome feedback from other reviewers.

Reply:
Thanks @sdasgup3 and @sngyhan  for your reply.

I think the first point is resolved by the comment from @loganchien who points out that zero-point range must be restricted to storage range so the value 0 can be represented. From this I understand that zero-point will be limited to storage range?

On the second point, I’d like to expand a bit more the problems of having zero points on larger data types (greater than 8 bit). For example, an si16 data type with an si16 zero point can give data range of -65535 to +65535. Multiplying two such values will overflow a 32-bit result type. This then has an implementation cost. Since zero_point seems most effective on small data types which have limited range and precision, I’d suggest not to allow on larger data types unless there is a clear use case. I think also in TFLite the zero-point value for integer 16-bit quantized type is set to zero in usage.

Reply:
>> Since zero_point seems most effective on small data types which have limited range and precision, I’d suggest not to allow on larger data types unless there is a clear use case.

@dominicsymes I can see you point on restricting the type of zero point to 8-bits. Thanks for explaining!

@loganchien @sngyhan Do you mind providing your thoughts here?  

For unsigned quantized types (> 8-bits), the zero point  cannot be zero and need to match the storage type. Can this be a use-case?

Reply:
I prefer not adding such constraint at this moment.  IMO, StableHLO is more of the compiler IR that aims to support different targets instead of a least common denominator that everyone must support.  So, for this particular case, I prefer to keep the symmetry of IR design rather than adding a constraint just because some targets can't efficiently implement it.

I understand that spec has bidirectional effects, i.e. the implementors want to know what are the features must be implemented and the users (quantizers) want to know what are the features that can be used.  But, I think we can defer that later into the concept of profiles or compatibility matrix.

For concrete usages, I only find a usage when the signed integer type is emulated with the quantized type.  However, that is only an internal usage and can be implemented differently (thus does not necessary coupling with StableHLO spec).  As of writing, I don't find quantize models with >8bit zero point.  Maybe @sngyhan will have a better idea because quantizer is not my full-day work.

Reply:
IMO, an authoring tool/framework doesn't have to use a feature just because the spec can support/express such feature.  The authoring tool (producer) can still limit themselves to use `zp in [-128, 127]` for `i8`/`i16` if the narrower zero point range can accommodate their need.  Interpreters/validators may have a valid case because they are on the implementor side (consumer).  But, like the other parts of the spec, I don't think an interpreter can efficiently cover all features.  And, we will still have per-op constraints which can be used to narrow down the scope as needed (all multiplicative ops may need a narrower range, which I agree).  For visualization, I don't think it will add constraints to the value range.

My point is not to enact a constraint too early.  Maybe someday we will evolve to a stage with real need in real model.

My proposed wording is to limit zero point to storage type, i.e. `i8` gets `[-128, 127]`, `u8` gets `[0, 255]`, `i16` gets `[-32768, 32767]`, etc.  Thus, only `i64` storage type can get `i64` zero point.

But I am fine with adding constraints (albeit not preferring).  Actually, the implementation I am working on probably can't support all zero points in all operations either (in case the product overflows as pointed out by @dominicsymes earlier).  My comments is more on the design symmetry or my personal view on aesthetic.

Reply:
Thanks everybody for the comments! 
@sngyhan Can you please add your opinion as well.

Reply:
I am also not preferring specific constraints based on we've observed from TFLite product. IMO, it potentially might limit the adoptions of new quantization schemes/kernels in the future, e.g., experimental quantization schemes.

Numerically, to cover all the possible cases, "original value range ⊂ zp range" might be the right constraint.

Reply:
This is a very lively discussion! Given the variety of proposals, I think that this conversation would benefit from a video meeting. I'll work with @theadactyl to set up a community meeting about quantization, and this will be one of the topics. We've opened #1405 to keep track.